src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -99,6 +101,66 @@ protected override Stream ResponseStream
             get { return this.responseStream; }
         }
 
+        /// <summary>
+        /// Asynchronous bulk update request.
+        /// </summary>
+        /// <typeparam name="T"> The type of the top-level objects to be deep-updated.</typeparam>
+        /// <param name="objects"> The top-level objects of the type to be deep updated.</param>
+        internal void BeginBulkUpdateRequest<T>(params T[] objects)
+        {
+            PerRequest pereq = null;
+            ODataRequestMessageWrapper bulkUpdateRequestMessage = null;
+
+            if (objects == null || objects.Length == 0)
+            {
+                throw Error.Argument(Strings.Util_EmptyArray, nameof(objects));
+            }
+
+            BuildDescriptorGraph(this.ChangedEntries, true, objects);
+            
+            try 
+            {
+                bulkUpdateRequestMessage = this.GenerateBulkUpdateRequest();
+                this.Abortable = bulkUpdateRequestMessage;
+
+                if (bulkUpdateRequestMessage != null)
+                {
+                    bulkUpdateRequestMessage.SetContentLengthHeader();
+                    this.perRequest = pereq = new PerRequest();
+                    pereq.Request = bulkUpdateRequestMessage;
+                    pereq.RequestContentStream = bulkUpdateRequestMessage.CachedRequestStream;
+
+                    AsyncStateBag asyncStateBag = new AsyncStateBag(pereq);
+
+                    this.responseStream = new MemoryStream();
+
+                    IAsyncResult asyncResult = BaseAsyncResult.InvokeAsync(bulkUpdateRequestMessage.BeginGetRequestStream, this.AsyncEndGetRequestStream, asyncStateBag);
+                    
+                    pereq.SetRequestCompletedSynchronously(asyncResult.CompletedSynchronously);
+                }
+                else 
+                {
+                    this.SetCompleted();
+
+                    if (this.CompletedSynchronously)
+                    {
+                        this.HandleCompleted(pereq);
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                this.HandleFailure(pereq, e);
+                throw; 
+            }
+            finally
+            {
+                this.HandleCompleted(pereq);

```

Comment:
Yes we dispose the request object.

Reply:
Where is it getting disposed?

I see that we're disposing `perReq` at some point, but looking at the source code of `PerRequest.Dispose()`, it doesn't seem to be disposing it's `PerRequest.Request` object. So it's still not clear to me at what point we're disposing `this.bulkUpdateRequestMessage`.

Reply:
`perReq` is a wrapper of the request object. When it is disposed it disposes its contents.

Reply:
I see. It's strange that it doesn't implement `IDisposable`, seems like design flaw. I wonder whether we may have any resource leaks since at least the underlying `HttpClientRequestMessage` is disposable.